### PR TITLE
[native-pos] Disable AsyncDataCache (thus prefetching)

### PR DESCRIPTION
**Pull request by Yuhta**
We see negative performance impact (5% ~ 30%) when split prefetch is enabled,
hence adding these config options to disable it.

### CODE REVIEW

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java [hunk 1]
```diff
@@ -323,6 +325,18 @@ public int getNumIoThreads()
...
     }
 
+    @Config(NUM_CONNECTOR_IO_THREADS)
+    public NativeExecutionSystemConfig setNumConnectorIoThreads(int numConnectorIoThreads)
+    {
+        this.numConnectorIoThreads = numConnectorIoThreads;
+        return this;
+    }
+
+    public int getNumConnectorIoThreads()
```
Comment (shrinidhijoshi) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java:335]: Also add a call to this function in the `getAllProperties` function on L127.

**Comment by shrinidhijoshi** [comment]
@Yuhta Please update the commit message (and PR description) with brief explanation of why this change is needed.

**Review by shrinidhijoshi — APPROVED** [review]
We tried to do a test offline but the internal tooling was slowing us down. As this is non-breaking and very specific to cpp process, I am approving this. We can come back and fix if something is missing

**Review by mbasmanova — COMMENTED** [review]
@Yuhta Use [native-pos] prefix for Presto-on-Spark specific changes.

**Review by mbasmanova — COMMENTED** [review]
@Yuhta What is the definition of num-connector-io-threads property? Is it documented already? If not, would you document it?

**Review by mbasmanova — COMMENTED** [review]
> We see negative performance impact (5% ~ 30%) when split prefetch is enabled,

What is causing that?

**Review by mbasmanova — COMMENTED** [review]
@Yuhta Since you have [native-pos] prefix, adding "for Presto on Spark" is redundant. Let's remove from both PR title and commit message.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java [hunk 2]
```diff
@@ -59,6 +59,9 @@ public class NativeExecutionSystemConfig
     // TODO: others use "-" separator and this property use _ separator. Fix them.
     private static final String HTTP_EXEC_THREADS = "http_exec_threads";
     private static final String NUM_IO_THREADS = "num-io-threads";
+    private static final String ASYNC_DATA_CACHE_ENABLED = "async-data-cache-enabled";
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java:62]: Do we have these properties documented? If not, please, document them.

**Review by mbasmanova — COMMENTED** [review]
> We see negative performance impact (5% ~ 30%) when split prefetch is enabled,

@Yuhta What's the cause of performance impact?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java (hunk 2 above)
Comment (Yuhta) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java:62]: Documented in where they are defined

Comment (Yuhta) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java:62]: Where should I put it?

File: presto-docs/src/main/sphinx/develop/presto-native.rst [hunk 3]
```diff
@@ -126,4 +126,34 @@ There is also an additional parameter:
...
+--------------------------------
+
+``num-connector-io-threads``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type** ``integer``
+    * **Default value:** ``30``
+
+    Size of IO executor for connectors to do preload/prefetch.  Prefetch is
+    disabled if set to zero.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/develop/presto-native.rst:141]:
```suggestion
    disabled if ``num-connector-io-threads`` is set to zero.
```
Suggest explicitly stating what to set to 0 to disable prefetch.

File: presto-docs/src/main/sphinx/develop/presto-native.rst [hunk 4]
```diff
@@ -126,4 +126,34 @@ There is also an additional parameter:
...
+
+``async-data-cache-enabled``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type** ``bool``
+    * **Default value:** ``true``
+
+    Whether async data cache is enabled.  Disable this also disable split
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/develop/presto-native.rst:149]:
```suggestion
    Whether async data cache is enabled.  Setting ``async-data-cache-enabled`` to ``false`` disables split
```
Suggesting specifying what is done and its result.

File: presto-docs/src/main/sphinx/develop/presto-native.rst [hunk 5]
```diff
@@ -126,4 +126,34 @@ There is also an additional parameter:
...
+    prefetching in table scan.
+
+``async-cache-ssd-gb``
+^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type** ``integer``
+    * **Default value:** ``0``
+
+    Size of the SSD cache when async data cache is enabled.  Must be zero if
+    async data cache is disabled.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/develop/presto-native.rst:159]:
```suggestion
    ``async-data-cache-enabled`` is disabled.
```

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the documentation! I added a few suggestions to make some statements more explicit and easier to understand. If I have made a suggestion that changes the meaning in a way you disagree with, please help me understand the meaning.

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

File: presto-docs/src/main/sphinx/develop/presto-native.rst [hunk 6]
```diff
@@ -126,4 +126,34 @@ There is also an additional parameter:
...
+    * **Default value:** ``true``
+
+    Whether async data cache is enabled.  Setting ``async-data-cache-enabled``
+    to ``false`` disables split prefetching in table scan.
+
+``async-cache-ssd-gb``
+^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type** ``integer``
+    * **Default value:** ``0``
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/develop/presto-native.rst:156]: Are we missing some other configs re: SSD. I assume one needs to specify rott directory or something, no?

File: presto-docs/src/main/sphinx/develop/presto-native.rst [hunk 7]
```diff
@@ -126,4 +126,34 @@ There is also an additional parameter:
...
+``async-data-cache-enabled``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type** ``bool``
+    * **Default value:** ``true``
+
+    Whether async data cache is enabled.  Setting ``async-data-cache-enabled``
+    to ``false`` disables split prefetching in table scan.
+
+``async-cache-ssd-gb``
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/develop/presto-native.rst:152]: consistency: async-data-cache-enabled vs. async-cache-ssd-gb - one says data-cache, the other just cache. Would be nice to fix.

File: presto-docs/src/main/sphinx/develop/presto-native.rst [hunk 8]
```diff
@@ -126,4 +126,34 @@ There is also an additional parameter:
...
     is valid.
     If the time period exceeds the parameter value, the request is rejected as
-    authentication failure (HTTP 401).
\ No newline at end of file
+    authentication failure (HTTP 401).
+
+Async Data Cache and Prefetching
+--------------------------------
+
+``num-connector-io-threads``
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/develop/presto-native.rst:134]: Is this specific to Hive connector?

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java [hunk 9]
```diff
@@ -104,6 +107,9 @@ public class NativeExecutionSystemConfig
     private String httpsCertPath = "";
     private String httpsKeyPath = "";
     private int numIoThreads = 30;
+    private boolean asyncDataCacheEnabled;
```
Comment (mbasmanova) [inline presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java:110]: Should these have default values? true, 0 and 30 (according to documentation).

File: presto-docs/src/main/sphinx/develop/presto-native.rst (hunk 7 above)
Comment (Yuhta) [reply presto-docs/src/main/sphinx/develop/presto-native.rst:152]: That will be a big change, rather not changing it here

Comment (mbasmanova) [reply presto-docs/src/main/sphinx/develop/presto-native.rst:152]: Understood. Perhaps, file a GitHub issue about this.

File: presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java (hunk 9 above)
Comment (Yuhta) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java:110]: We set it to false here to disable it.  The default values are for when it is not overwritten here.

Comment (mbasmanova) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java:110]: Would you explicitly set it or make a comment that default is 'false'? 

```
private boolean asyncDataCacheEnabled = false;
```

Comment (mbasmanova) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java:110]: > The default values are for when it is not overwritten here.

I'm not sure I understand.... The user reading the documentation would assume that default is what they get without changing anything. It looks like actual default is false, no?

Comment (Yuhta) [reply presto-spark-base/src/main/java/com/facebook/presto/spark/execution/property/NativeExecutionSystemConfig.java:110]: Setting it explicitly violate a lint rule.  I will put a comment.

The default is you don't overwrite it in application code.  Here POS is overwriting it explicitly so the defaults no longer applies.  Do you have a place I can write POS defaults?

File: presto-docs/src/main/sphinx/develop/presto-native.rst (hunk 6 above)
Comment (Yuhta) [reply presto-docs/src/main/sphinx/develop/presto-native.rst:156]: Just you will need to set SSD size in another property to indicate how much space is available

File: presto-docs/src/main/sphinx/develop/presto-native.rst (hunk 7 above)
Comment (Yuhta) [reply presto-docs/src/main/sphinx/develop/presto-native.rst:152]: #21185

**Review by mbasmanova — APPROVED** [review]
Accepting to unblock, but I'm very confused by how these properties are set and what the defaults are. It would be nice to clear this up in a follow up.

**Comment by aditi-pandit** [comment]
@Yuhta , @mbasmanova : We noticed this negative performance impact internally for Prestissimo as well. Did you'll only see it for POS, but not Prestissimo ? (Just curious as you are only turning off POS configs)

We should turn it off in https://github.com/prestodb/presto/blob/master/presto-native-execution/src/test/java/com/facebook/presto/nativeworker/NativeQueryRunnerUtils.java#L33 config as well. wdyt ?

**Comment by Yuhta** [comment]
@aditi-pandit We have different deployments of Prestissimo and some do turn the option off.  You can use the worker config file to overwrite the system config.  POS needs special treatment because the configuration is a little bit weird and worker config does not work out of box there.

**Pull request merged**